modules/dnn/src/graph_simplifier.cpp

```diff
@@ -110,13 +111,15 @@ bool Subgraph::match(const Ptr<ImportGraphWrapper>& net, int nodeId,
             if (nodes[inputNodes[j]].empty())  // Unknown input node type.
                 continue;
             nodeId = getInputNodeId(net, node, j);
+            if (nodeId == -1)
+                return false;
```

Comment:
That's strange. this method trying to find `node.input[j]` in the network `net`. When it can cannot find a node?

Reply:
![Screenshot from 2020-02-20 07-48-37](https://user-images.githubusercontent.com/26292819/74894639-9871c400-53b5-11ea-8e86-bbf19f123c6b.png)

```
names of all nodes are :
name - tensor   Type - 
name - 1        Type - Split. 
name - 3        Type - Concat 
```
And the input names of `Concat` are 1, 2 and there is no node with name 2.

Since `ExtractScalesSubgraph` fuse this subgraph to node `Concat` ang in split test subgraph Concat layer is present, it will try to match with 'Concat' layer of `ExtractScalesSubgraph`.
![Screenshot from 2020-02-16 19-00-03](https://user-images.githubusercontent.com/26292819/74894853-20f06480-53b6-11ea-934a-89beb6be7a31.png)
And there is no node with name `2` in split test subgraph. Split layer has output `1` and `2` but it is named as `1`. 

Before program was stopping at line 75.
https://github.com/opencv/opencv/blob/4b0d3316f6e1437c1fee98934f6036e96d1d321b/modules/dnn/src/graph_simplifier.cpp#L70-L75

but in my view it shouldn't stop. It must simply move on. 
https://github.com/opencv/opencv/blob/23ab3cd59045c0d583a9f233b2f3885184101c9a/modules/dnn/src/graph_simplifier.cpp#L70-L76
https://github.com/opencv/opencv/blob/23ab3cd59045c0d583a9f233b2f3885184101c9a/modules/dnn/src/graph_simplifier.cpp#L113-L115

So, in conclusion, this problem would arise if some node giving more than one output to any layer taking those as inputs. just in case of `split` and `concat`  in split test subgraph. 
Pls correct me if i am wrong.

Reply:
Can you please specify what is Split test? Looks like in this PR there is no subgraphs with Split layer?

Reply:
ohh sorry. I am talking about these tests. 
https://github.com/opencv/opencv/blob/4d14307f5d4aaabe9af29e72c17102cce50d883a/modules/dnn/test/test_onnx_importer.cpp#L392-L401
and the image shared is of split_1.
![Screenshot from 2020-02-20 07-48-37](https://user-images.githubusercontent.com/26292819/74932013-dcd98000-5406-11ea-86e1-907dd09d1f8f.png)

Reply:
These test were failing because of above mentioned reason.

Reply:
Got it, thanks!
So maybe we can modify `getInputNodeId` in this way?

Due we build patterns without output indices, so we can return a node which produces one of the outputs.

In this example with Split,

```
         0
       +--->
split -+     concat
       +--->
         1
```

`getInputNodeId(net, concatNode, 0) -> Split`
and
`getInputNodeId(net, concatNode, 1) -> Split`

So this method checks all the output names for nodes and if it matches - return it.

Reply:
But there in no function like `getNumOutputs()` or `getOutputName()` in `class ImportNodeWrapper
` so I think to get output names we need to add these two functions in `class ImportNodeWrapper` ?

Reply:
Agree, `int getNumOutputs()` and add an index for `getOutputName(int idx = 0)`

Reply:
@dkurt , is it fine to add 
```
    virtual int getNumOutputs() const
    {
        return 0;
    }
    virtual std::string getOutputName(int idx = 0) const
    {
        return "";
    }
```
instead of 

   ```
virtual int getNumOutputs() const = 0;
virtual std::string getOutputName(int idx) const = 0;
```
in graph_simplifier.hpp. 

Otherwise, we need to modify `class TFNodeWrapper : public ImportNodeWrapper`
in tf_graph_simplifier.cpp